### PR TITLE
kill `isleaftype`

### LINKED ISSUE #17086: Rename `isleaftype`
While re-reading the manual I realized that `isleaftype` should probably be renamed to either `isconcretetype` or `isconcrete`:

``` julia
julia> abstract Foo

julia> abstract Bar <: Foo

julia> type Baz <: Foo
       end

julia> isleaftype(Bar)  # One could consider Bar a leaf type but not concrete
false

julia> isleaftype(Baz)
true
```

The documentation for `isleaftype` does clearly explain it returns true for concrete types which seems like the function name is slightly misleading.

### LINKED ISSUE #23567: Are isbits-Unions isbits?
Noticed in the current master bug:
```julia
julia> t = Vector{Union{Float64, Void}}(5)
5-element Array{Union{Void, Float64},1}:
 nothing
 nothing
 nothing
 nothing
 nothing

julia> b = collect(Union{Float64, Void}, 1.0:3.0)
3-element Array{Union{Void, Float64},1}:
 1.0
 2.0
 3.0

julia> copy!(t, 2, b)
5-element Array{Union{Void, Float64},1}:
 0.0
 1.0
 2.0
  nothing
  nothing
```
The issue is that we currently go thru the codepath in array.jl
```julia
function unsafe_copy!(dest::Array{T}, doffs, src::Array{T}, soffs, n) where T
    if isbits(T)
        unsafe_copy!(pointer(dest, doffs), pointer(src, soffs), n)
    else
        ccall(:jl_array_ptr_copy, Void, (Any, Ptr{Void}, Any, Ptr{Void}, Int),
              dest, pointer(dest, doffs), src, pointer(src, soffs), n)
    end
```
but `isbits(Union{Float64, Void})` is false; so we hit `jl_array_ptr_copy`. Now, I did had a hook in `jl_array_ptr_copy` to intercept isbits-Union arrays and handle them specially, the problem is it currently only works when you happen to be copying to the 1st element of the dest array. This is because we only pass `dest_p`, a pointer and not an index, to `jl_array_ptr_copy`, so we don't actually know where to copy selector bytes from src to dest.

There a couple of ways to address this:
- go ahead and make `isbits(Union{Float64, Void}) == true`, this would involve making sure we're careful in a # of places in Base where we're currently checking `isbits(T)` on arrays and doing optimizations (i.e. `vcat`, `unsafe_copy`, `flipdim`). The biggest ? for me in doing this is how this propagates thru the rest of the system, in particular, inference.jl (which has quite a few calls to `isbits(T)`.). Overall, I think it would be more correct, but it does seem a bit worrisome not quite knowing the full extent of this kind of change. I can certainly do more research here if we think it's the best way to try out first.
- make another method `jl_array_ptr_copy2`, which also takes `ind` arguments so we can correctly adjust when copying isbits-Union arrays.

Ultimately, I think we _should_ make isbits-Unions isbits, because we're essentially doing that everywhere in `src/`, so we might as well be consistent in Base.

cc: @JeffBezanson, @vtjnash

### LINKED ISSUE #25044: fix isbits predicate
Right now, it returns wrong answers. Also, we should look into fixing the `isleaftype` predicate.

### LINKED ISSUE #25310: EXCEPTION_ACCESS_VIOLATION for a small example
````
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-DEV.3095 (2017-12-19 01:41 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 10443357f6* (9 days old master)
|__/                   |  x86_64-w64-mingw32

julia> struct Nullable2{T}
         val::Union{Some{T},Void}
       end

julia> dest_nullable2 = Array{Nullable2{Int},1}(uninitialized, 10)
10-element Array{Nullable2{Int64},1}:

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x6b5ccc94 -- jl_get_nth_field at /home/Administrator/buildbot/worker/package_win64/build/src\datatype.c:803
in expression starting at no file:0
jl_get_nth_field at /home/Administrator/buildbot/worker/package_win64/build/src\datatype.c:803
jl_f_getfield at /home/Administrator/buildbot/worker/package_win64/build/src\builtins.c:753
show_default at .\show.jl:153
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
show at .\show.jl:136
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
jl_apply at /home/Administrator/buildbot/worker/package_win64/build/src\julia.h:1474 [inlined]
jl_f__apply at /home/Administrator/buildbot/worker/package_win64/build/src\builtins.c:556
#sprint#285 at .\strings\io.jl:89
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011 [inlined]
jl_apply at /home/Administrator/buildbot/worker/package_win64/build/src\julia.h:1474 [inlined]
jl_invoke at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:51
#sprint at .\<missing>:0 [inlined]
alignment at .\show.jl:1624 [inlined]
alignment at .\arrayshow.jl:68
unknown function (ip: 00000000140F7809)
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
print_matrix at .\arrayshow.jl:187
print_matrix at .\arrayshow.jl:160 [inlined]
print_array at .\arrayshow.jl:309 [inlined]
_display at .\arrayshow.jl:345
show at .\replutil.jl:139
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
display at .\repl\REPL.jl:126
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
display at .\repl\REPL.jl:129
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
display at .\multimedia.jl:291
inner at .\essentials.jl:665
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
jl_apply at /home/Administrator/buildbot/worker/package_win64/build/src\julia.h:1474 [inlined]
jl_f__apply at /home/Administrator/buildbot/worker/package_win64/build/src\builtins.c:556
jl_f__apply_latest at /home/Administrator/buildbot/worker/package_win64/build/src\builtins.c:594
#invokelatest#3 at .\essentials.jl:666 [inlined]
invokelatest at .\essentials.jl:665 [inlined]
print_response at .\repl\REPL.jl:147
unknown function (ip: 00000000140F140D)
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
print_response at .\repl\REPL.jl:133
unknown function (ip: 00000000140F0F3D)
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
do_respond at .\repl\REPL.jl:708
unknown function (ip: 00000000140F0896)
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
do_call at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:323
eval_value at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:395
eval_body at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:505
jl_interpret_toplevel_thunk_callback at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:720
unknown function (ip: FFFFFFFFFFFFFFFE)
unknown function (ip: 000000001000E84F)
unknown function (ip: FFFFFFFFFFFFFFFF)
jl_toplevel_eval_flex at /home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:721
jl_toplevel_eval_in at /home/Administrator/buildbot/worker/package_win64/build/src\builtins.c:626
eval at .\boot.jl:289 [inlined]
eval at .\repl\LineEdit.jl:3
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
run_interface at .\repl\LineEdit.jl:2228
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
run_frontend at .\repl\REPL.jl:1024
run_repl at .\repl\REPL.jl:183
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
_start at .\client.jl:420
jl_call_fptr_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:380 [inlined]
jl_call_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\julia_internal.h:399 [inlined]
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2011
jl_apply at /home/Administrator/buildbot/worker/package_win64/build/src\julia.h:1474 [inlined]
true_main at /home/Administrator/buildbot/worker/package_win64/build/ui\repl.c:107
wmain at /home/Administrator/buildbot/worker/package_win64/build/ui\repl.c:237
__tmainCRTStartup at /usr/src/debug/mingw64-x86_64-runtime-5.0.2-1/crt\crtexe.c:329
mainCRTStartup at /usr/src/debug/mingw64-x86_64-runtime-5.0.2-1/crt\crtexe.c:212
BaseThreadInitThunk at C:\WINDOWS\System32\KERNEL32.DLL (unknown line)
RtlUserThreadStart at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
Allocations: 5170144 (Pool: 5167427; Big: 2717); GC: 11
````

### PR DESCRIPTION
**Pull request by vtjnash**
As discussed in previous issues (listed below), this finishes the deprecation of `isleaftype` and adds/corrects a number of other type-property queries and their usages (esp. including `isbits`). This correction also includes a number of changes that makes inference more powerful (by leveraging the removal of isleaftype to return more correct answers, yay!).

Most of the new computations are cached inside a datatype upon construction:
```c
// properties of `parameters`
     uint8_t hasfreetypevars; // majority part of isconcrete computation
+    uint8_t isconcretetype; // where this type can have instances
+    uint8_t isdispatchtuple; // aka isleaftupletype
// properties of `types`
+    uint8_t isbitstype; // relevant query for C-api and type-parameters
+    uint8_t zeroinit; // if one or more fields requires zero-initialization
+    uint8_t isinlinealloc; // if this is allocated inline
// removed attributes
-    int32_t depth; // not a well-defined concept
-    int8_t isleaftype; // already deprecated, now gone
```

fix #25044
fix #23567
fix #17086
fix #25310

~~This still needs docs from https://github.com/JuliaLang/julia/issues/17086#issuecomment-328912370, and I should go back through and add some tests. But wanted to start by publish the functional changes now.~~

### CODE REVIEW

File: src/intrinsics.cpp [hunk 1]
```diff
@@ -297,12 +310,8 @@ static Value *emit_unboxed_coercion(jl_codectx_t &ctx, Type *to, Value *unboxed)
             unboxed = ctx.builder.CreateBitCast(unboxed, INTT_to);
         unboxed = ctx.builder.CreateIntToPtr(unboxed, to);
     }
-    else if (ty == T_int1 && to == T_int8) {
-        // bools may be stored internally as int8
-        unboxed = ctx.builder.CreateZExt(unboxed, T_int8);
-    }
     else if (ty != to) {
-        unboxed = ctx.builder.CreateBitCast(unboxed, to);
+            unboxed = ctx.builder.CreateBitCast(unboxed, to);
```
Comment (JeffBezanson) [inline src/intrinsics.cpp]: indentation

**Comment by JeffBezanson** [comment]
:+1: Looks great!

File: base/reflection.jl [hunk 2]
```diff
@@ -296,65 +317,60 @@ julia> isbits(Complex)
...
 """
-isbits(t::DataType) = (@_pure_meta; !t.mutable & (t.layout != C_NULL) && datatype_pointerfree(t))
-isbits(t::Type) = (@_pure_meta; false)
-isbits(x) = (@_pure_meta; isbits(typeof(x)))
+isbits(@nospecialize(t::Type)) = (@_pure_meta; isa(t, DataType) && t.isbitstype)
+isbits(@nospecialize x) = (@_pure_meta; typeof(x).isbitstype)
 
-_isleaftype(@nospecialize(t)) = (@_pure_meta; isa(t, DataType) && t.isleaftype)
+_isleaftype(@nospecialize(t)) = (@_pure_meta; isa(t, DataType) && (t.isdispatchtuple || t.isconcretetype)) # deprecated
+isdispatchtuple(@nospecialize(t)) = (@_pure_meta; isa(t, DataType) && t.isdispatchtuple)
```
Comment (timholy) [inline base/reflection.jl]: Docs? It's exported.

File: base/reflection.jl [hunk 3]
```diff
@@ -277,15 +321,45 @@ false
...
+function isstructtype(@nospecialize(t::Type))
+    @_pure_meta
+    t = unwrap_unionall(t)
+    # TODO: what to do for `Union`?
+    isa(t, DataType) || return false
+    return length(t.types) != 0 || (t.size == 0 && !t.abstract)
+end
+
+"""
+    Base.isstructype(T) -> Bool
```
Comment (nalimilan) [inline base/reflection.jl]: `isprimitivetype`

File: base/reflection.jl [hunk 4]
```diff
@@ -277,15 +321,45 @@ false
 """
 isimmutable(@nospecialize(x)) = (@_pure_meta; !typeof(x).mutable)
 
-isstructtype(t::DataType) = (@_pure_meta; length(t.types) != 0 || (t.size==0 && !t.abstract))
-isstructtype(x) = (@_pure_meta; false)
+"""
+    Base.isstructype(T) -> Bool
```
Comment (nalimilan) [inline base/reflection.jl]: `isstructtype`

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/25496#issuecomment-357776987):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/7bfa97bfc9639739d1bc408e59c362974b3155f9/e16aedd_vs_5e2ff12/report.md)
cc @ararslan

**Comment by ararslan** [comment]
Random moving to the stdlib broke Nanosoldier. I have a fix up, I just need to retune the benchmarks then I'll run Nanosoldier here when that's done.

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25496#issuecomment-357849698) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/ff840fbd441840c4caa9b90fb8f2c92c8a064986/e727c3d_vs_5e2ff12/report.md). cc @ararslan

**Comment by vtjnash** [comment]
Seems mostly noise, but let me run that again to see if any are consistently changed: @nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25496#issuecomment-358087762) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6400bd9ecaa3e2fb4f9ec13d149044873c6c8aa9/741ccfa_vs_aea9155/report.md). cc @ararslan

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25496#issuecomment-358207644) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/aa12c64694f502915032531c98bb950e95f349d0/c5a9e52_vs_d6f079a/report.md). cc @ararslan

**Comment by JeffBezanson** [comment]
:tada:

**Pull request merged**